api/src/main/java/io/cloudevents/rw/CloudEventDataMapper.java

```diff
@@ -37,4 +37,10 @@
      */
     R map(CloudEventData data) throws CloudEventRWException;
 
+    /**
+     * No-op identity mapper which can be used as default when no mapper is provided.
+     */
+    static CloudEventDataMapper<? extends CloudEventData> identity() {
```

Comment:
Yeah, I hadn't noticed the problem with the type bound. I guess this is fine. The bound `extends CloudEventData` is redundant because it's already present in the declaration of `CloudEventDataMapper`. In other words just `<?>` would have the same effect.

If the wildcard does cause problems it's possible to restore the `<R>` solution using casts:

```java
    @SuppressWarnings("unchecked")
    static <R extends CloudEventData> CloudEventDataMapper<R> identity() {
      return d -> (R) d;
    }
```

Reply:
By the way this suggests to me that the types here aren't really sound. This is saying that you can take any CloudEventData and convert it to MyCloudEventData with an identity transformation. I wonder if the `<R>` parameter is really useful.

Reply:
> I wonder if the <R> parameter is really useful.

I can show you a use case when this is useful: https://github.com/cloudevents/sdk-java/pull/288/files#diff-d32fb9f764f16209b2cb69f1239b4309975c5601affe825cc4ff9428eca42d1bR52

The `CloudEventDataMapper` can be used directly by the user, when interacting with the event. In this use case, invoking the mapper helper, you get back the concrete type extending `CloudEventData`. Or you can use the mapper inside a `CloudEventReader` or `EventFormat`, and you won't care about the returned concrete type at all because in any case, you assign back to a `CloudEventData`.

Reply:
Maybe in the identity case `R` should just be `CloudEventData`? like:

```java
    static CloudEventDataMapper<CloudEventData> identity() {
      return d -> d;
    }
```

Reply:
Yes! That's obviously the right thing to do. Sorry I made it more complicated.

Reply:
ok can you review it again? If it's fine, i'm gonna merge it :smile: